File: JavaLibrary/src/main/java/com/flat/simplex/lexer/Lexer.java
package com.flat.simplex.lexer;

public class Lexer {

    private int chr;
    private int nChr;
    private int index;
    private int nextIndex;
    private final String source;

    public Lexer(String source) {
        this.source = source;
        readNextChar();
    }

    public Token read() {
        Token start = readNext();
        Token token = start;
        while (token != null) {
            token.setNext(readNext());
            token = token.getNext();
        }
        return start;
    }

    private boolean eof() {
        return index >= source.length();
    }

    private Token readNext() {
        while (!eof()) {
            if (isSpace(chr)) {
                consumeSpaces();
                continue;

            } else if (chr == '/' && nChr == '/') {
                consumeNextLineComment();
                continue;

            } else if (chr == '/' && nChr == '*') {
                consumeNextBlockComment();
                continue;

            }

            if (isOpener(chr)) {
                return readNextBlock();

            } else if (isCloser(chr)) {
                return readNextCloser();

            } else if (isLetter(chr)) {
                return readNextWord();

            } else if (isNumber(chr) || chr == '#') {
                return readNextNumber();

            } else if (isQuot(chr)) {
                return readNextString();

            } else if (isSplitter(chr)) {
                return readNextSplitter();

            } else if (isOperator(chr)) {
                return readNextOperator();

            } else {
                return readInvalidChars();
            }
        }
        return null;
    }

    private void readNextChar() {
        index = nextIndex;
        if (index >= source.length()) {
            chr = 0;
            nChr = 0;
        } else {
            chr = source.codePointAt(index);
            nextIndex = index + Character.charCount(chr);
            nChr = nextIndex < source.length() ? source.codePointAt(nextIndex) : 0;
        }
    }

    private void consumeSpaces() {
        while (!eof() && isSpace(chr)) {
            readNextChar();
        }
    }

    private void consumeNextLineComment() {
        readNextChar();
        readNextChar();
        while (!eof()) {
            if (chr == '\n') {
                readNextChar();
                return;
            }
            readNextChar();
        }
    }

    private void consumeNextBlockComment() {
        readNextChar();
        readNextChar();
        while (!eof()) {
            if (chr == '*' && nChr == '/') {
                readNextChar();
                readNextChar();
                return;
            }
            readNextChar();
        }
    }

    private boolean isCloser(Key a, Key b) {
        if (a == Key.Param) return b == Key.CParam;
        if (a == Key.Brace) return b == Key.CBrace;
        if (a == Key.Index) return b == Key.CIndex;
        return false;
    }

    private Token readNextBlock() {
        int s = index;
        readNextChar();
        Key key = Key.readKey(source, s, index);
        Token block = new Token(source, s, index, key == null ? Key.Invalid : key);

        Token next = readNext();
        Token start = null;
        Token token = null;
        while (next != null) {

            if (start == null) {
                start = next;
            } else {
                token.setNext(next);
            }
            token = next;
            if (isCloser(key, token.getKey())) {
                block.setLastChild(token);
                break;
            }
            next = readNext();
        }
        block.setChild(start);
        return block;
    }

    private Token readNextCloser() {
        int start = index;
        readNextChar();
        Key key = Key.readKey(source, start, index);
        return new Token(source, start, index, key == null ? Key.Invalid : key);
    }

    private Token readNextWord() {
        int start = index;
        while (!eof() && isChar(chr)) {
            readNextChar();
        }
        Key key = Key.readKey(source, start, index);
        return new Token(source, start, index, key == null ? Key.Word : key);
    }

    private Token readNextNumber() {
        int start = index;
        while (!eof()) {
            if (chr == 'e' && (nChr == '-' || nChr == '+')) {
                readNextChar();
                readNextChar();
            } else if (isLiteral(chr)) {
                readNextChar();
            } else {
                break;
            }
        }
        return new Token(source, start, index, Key.Number);
    }

    private Token readNextString() {
        int split = chr;
        int start = index;
        boolean invert = false;
        readNextChar();
        while (!eof()) {
            if (chr == '\\') {
                invert = !invert;
            } else if ((chr == split && !invert) || chr == '\n') {
                readNextChar();
                break;
            }

            readNextChar();
        }
        return new Token(source, start, index, Key.String);
    }

    private Token readNextSplitter() {
        int start = index;
        readNextChar();
        Key key = Key.readKey(source, start, index);
        return new Token(source, start, index, key == null ? Key.Invalid : key);
    }

    private Token readNextOperator() {
        int start = index;
        while (!eof() && isOperator(chr)) {
            readNextChar();
        }
        Key key = Key.readKey(source, start, index);
        return new Token(source, start, index, key == null ? Key.Invalid : key);
    }

    private Token readInvalidChars() {
        int start = index;
        while (!eof() && !isValidChar(chr)) {
            readNextChar();
        }

        return new Token(source, start, index, Key.Invalid);
    }

    private static boolean isSpace(int chr) {
        return Character.isSpaceChar(chr) || Character.isWhitespace(chr);
    }

    private static boolean isNumber(int chr) {
        return (chr >= '0' && chr <= '9');
    }

    private static boolean isLetter(int chr) {
        return (chr >= 'A' && chr <= 'Z') || (chr >= 'a' && chr <= 'z') || chr == '_';
    }

    private static boolean isOpener(int chr) {
        return chr == '{' || chr == '(' || chr == '[';
    }

    private static boolean isCloser(int chr) {
        return chr == '}' || chr == ')' || chr == ']';
    }

    private static boolean isOperator(int chr) {
        return chr == '+' || chr == '-' || chr == '*' || chr == '/' || chr == '%'
                || chr == '=' || chr == '!' || chr == '|' || chr == '&' || chr == '^' || chr == ':' || chr == '?'
                || chr == '~' || chr == '<' || chr == '>';
    }

    private static boolean isSplitter(int chr) {
        return chr == '.' || chr == ',' || chr == ';';
    }

    private static boolean isQuot(int chr) {
        return chr == '"' || chr == '\'';
    }

    private static boolean isValidChar(int chr) {
        return isSpace(chr) || isNumber(chr) || isLetter(chr) ||
                isOperator(chr) || isQuot(chr) || isSplitter(chr) || isOpener(chr) || isCloser(chr);
    }

    private static boolean isChar(int chr) {
        return isNumber(chr) || isLetter(chr);
    }

    private static boolean isLiteral(int chr) {
        return (chr >= '0' && chr <= '9') ||
               (chr >= 'A' && chr <= 'F') ||
               (chr >= 'a' && chr <= 'f') || chr == '.' || chr == '#';
    }
}

File: SimplexApp/simplex/lexer/Lexer.js
var Key, Token;

class Lexer {
    constructor(source) {
        this.chr = 0;
        this.nChr = 0;
        this.index = 0;
        this.nextIndex = 0;
        this.source = source;
        this.readNextChar();
    }

    read() {
        let start = this.readNext();
        let token = start;
        while (token !== null) {
            token.setNext(this.readNext());
            token = token.getNext();
        }
        return start;
    }

    eof() {
        return this.index >= this.source.length;
    }

    readNext() {
        while (!this.eof()) {
            if (Lexer.isSpace(this.chr)) {
                this.consumeSpaces();
                continue;

            } else if (this.chr === 47 && this.nChr === 47) {       //   (this.chr == '/' && this.nChr == '/')
                this.consumeNextLineComment();
                continue;

            } else if (this.chr === 47 && this.nChr === 42) {     //   (this.chr == '/' && this.nChr == '/')
                this.consumeNextBlockComment();
                continue;

            }

            if (Lexer.isOpener(this.chr)) {
                return this.readNextBlock();

            } else if (Lexer.isCloser(this.chr)) {
                return this.readNextCloser();

            } else if (Lexer.isLetter(this.chr)) {
                return this.readNextWord();

            } else if (Lexer.isNumber(this.chr) || this.chr === 35) {   // this.chr == '#'
                return this.readNextNumber();

            } else if (Lexer.isQuot(this.chr)) {
                return this.readNextString();

            } else if (Lexer.isSplitter(this.chr)) {
                return this.readNextSplitter();

            } else if (Lexer.isOperator(this.chr)) {
                return this.readNextOperator();

            } else {
                return this.readInvalidChars();
            }
        }
        return null;
    }

    readNextChar() {
        this.index = this.nextIndex;
        if (this.index >= this.source.length) {
            this.chr = 0;
            this.nChr = 0;
        } else {
            this.chr = this.source.codePointAt(this.index);
            this.nextIndex = this.index + (this.chr >= 0x10000 ? 2 : 1);
            this.nChr = this.nextIndex < this.source.length ? this.source.codePointAt(this.nextIndex) : 0;
        }
    }

    consumeSpaces() {
        while (!this.eof() && Lexer.isSpace(this.chr)) {
            this.readNextChar();
        }
    }

    consumeNextLineComment() {
        this.readNextChar();
        this.readNextChar();
        while (!this.eof()) {
            if (this.chr === 10) {      // (this.chr === '\n')
                this.readNextChar();
                return;
            }
            this.readNextChar();
        }
    }

    consumeNextBlockComment() {
        this.readNextChar();
        this.readNextChar();
        while (!this.eof()) {
            if (this.chr === 42 && this.nChr === 47) {      // (this.chr === '*' && this.nChr === '/')
                this.readNextChar();
                this.readNextChar();
                return;
            }
            this.readNextChar();
        }
    }

    isCloserFor(a, b) {
        if (a === Key.Param) {
            return b === Key.CParam;
        }
        if (a === Key.Brace) {
            return b === Key.CBrace;
        }
        if (a === Key.Index) {
            return b === Key.CIndex;
        }
        return false;
    }

    readNextBlock() {
        let s = this.index;
        this.readNextChar();
        let key = Key.readKey(this.source, s, this.index);
        let block = new Token(this.source, s, this.index, key === null ? Key.Invalid : key);
        let next = this.readNext();
        let start = null;
        let token = null;
        while (next !== null) {
            if (start === null) {
                start = next;
            } else {
                token.setNext(next);
            }
            token = next;
            if (this.isCloserFor(key, token.getKey())) {
                block.setLastChild(token);
                break;
            }
            next = this.readNext();
        }
        block.setChild(start);
        return block;
    }

    readNextCloser() {
        let start = this.index;
        this.readNextChar();
        let key = Key.readKey(this.source, start, this.index);
        return new Token(this.source, start, this.index, key === null ? Key.Invalid : key);
    }

    readNextWord() {
        let start = this.index;
        while (!this.eof() && Lexer.isChar(this.chr)) {
            this.readNextChar();
        }
        let key = Key.readKey(this.source, start, this.index);
        return new Token(this.source, start, this.index, key === null ? Key.Word : key);
    }

    readNextNumber() {
        let start = this.index;
        while (!this.eof()) {
            if (this.chr === 101 && (this.nChr === 45 || this.nChr === 43)) {
                this.readNextChar();
                this.readNextChar();
            } else {
                if (Lexer.isLiteral(this.chr)) {
                    this.readNextChar();
                } else {
                    break;
                }
            }
        }
        return new Token(this.source, start, this.index, Key.Number);
    }

    readNextString() {
        let split = this.chr;
        let start = this.index;
        let invert = false;
        this.readNextChar();
        while (!this.eof()) {
            if (this.chr === 92) {
                invert = !invert;
            } else {
                if ((this.chr === split && !invert) || this.chr === 10) {
                    this.readNextChar();
                    break;
                }
            }
            this.readNextChar();
        }
        return new Token(this.source, start, this.index, Key.String);
    }

    readNextSplitter() {
        let start = this.index;
        this.readNextChar();
        let key = Key.readKey(this.source, start, this.index);
        return new Token(this.source, start, this.index, key === null ? Key.Invalid : key);
    }

    readNextOperator() {
        let start = this.index;
        while (!this.eof() && Lexer.isOperator(this.chr)) {
            this.readNextChar();
        }
        let key = Key.readKey(this.source, start, this.index);
        return new Token(this.source, start, this.index, key === null ? Key.Invalid : key);
    }

    readInvalidChars() {
        let start = this.index;
        while (!this.eof() && !Lexer.isValidChar(this.chr)) {
            this.readNextChar();
        }
        return new Token(this.source, start, this.index, Key.Invalid);
    }

    static isSpace(chr) {
        return String.fromCharCode(chr).trim().length === 0;
    }

    static isNumber(chr) {
        // return (chr >= '0' && chr <= '9');
        return (chr >= 48 && chr <= 57);
    }

    static isLetter(chr) {
        // return (chr >= 'A' && chr <= 'Z') || (chr >= 'a' && chr <= 'z') || chr === '_';
        return (chr >= 65 && chr <= 90) || (chr >= 97 && chr <= 122) || chr === 95;
    }

    static isOpener(chr) {
        // return chr === '{' || chr === '(' || chr === '[';
        return chr === 123 || chr === 40 || chr === 91;
    }

    static isCloser (chr) {
        // return chr === '}' || chr === ')' || chr === ']';
        return chr === 125 || chr === 41 || chr === 93;
    }

    static isOperator(chr) {
        // return chr === '+' || chr === '-' || chr === '*' || chr === '/' || chr === '%'
        //     || chr === '=' || chr === '!' || chr === '|' || chr === '&' || chr === '^' || chr === ':' || chr === '?'
        //     || chr === '~' || chr === '<' || chr === '>';
        return chr === 43 || chr === 45 || chr === 42 || chr === 47 || chr === 37
            || chr === 61 || chr === 33 || chr === 124 || chr === 38 || chr === 94 || chr === 58 || chr === 63
            || chr === 126 || chr === 60 || chr === 62;
    }

    static isSplitter(chr) {
        // return chr === '.' || chr === ',' || chr === ';';
        return chr === 46 || chr === 44 || chr === 59;
    }

    static isQuot(chr) {
        // return chr === '"' || chr === '\'';
        return chr === 34 || chr === 39;
    }

    static isValidChar(chr) {
        return Lexer.isSpace(chr) || Lexer.isNumber(chr) || Lexer.isLetter(chr) ||
            Lexer.isOperator(chr) || Lexer.isQuot(chr) || Lexer.isSplitter(chr) ||
            Lexer.isOpener(chr) || Lexer.isCloser(chr);
    }

    static isChar(chr) {
        return Lexer.isNumber(chr) || Lexer.isLetter(chr);
    }

    static isLiteral(chr) {
        // return (chr >= '0' && chr <= '9') ||
        //     (chr >= 'A' && chr <= 'F') ||
        //     (chr >= 'a' && chr <= 'f') || chr === '.' || chr === '#';
        return (chr >= 48 && chr <= 57) ||
            (chr >= 65 && chr <= 70) ||
            (chr >= 97 && chr <= 102) || chr === 46 || chr === 35;
    }
}

module.exports = Lexer;
Key = require("simplex/lexer/Key.js");
Token = require("simplex/lexer/Token.js");
